packages/xchacha20poly1305/xchacha20poly1305.ts

```diff
@@ -0,0 +1,169 @@
+
+
+    // Use last 8 bytes of 24-byte extended nonce as an actual nonce prefixed by 4 NUL bytes,
+    // and a subkey derived in the previous step as key to encrypt.
+    const modifiedNonce = new Uint8Array(12);
+    modifiedNonce.set(nonce.subarray(16), 4);
+
+    const resultLength = plaintext.length + this.tagLength;
+    let result;
+    if (dst) {
+      if (dst.length !== resultLength) {
+        throw new Error(
+          "XChaCha20Poly1305: incorrect destination length"
+        );
+      }
+      result = dst;
+    } else {
+      result = new Uint8Array(resultLength);
+    }
+
+    const chachapoly1305 = new ChaCha20Poly1305(subKey);
+
+    chachapoly1305.seal(modifiedNonce, plaintext, associatedData, result);
+
+    return result;
+  }
+
+  /**
+   * Authenticates sealed ciphertext (which includes authentication tag) and
+   * associated data, decrypts ciphertext and returns decrypted plaintext.
+   *
+   * draft-irtf-cfrg-xchacha-01 defines a 24 byte nonce (192 bits) which
+   * then uses the first 16 bytes of the nonce and the secret key with
+   * Hchacha to generate an initial subkey. The last 8 bytes of the nonce
+   * are then prefixed with 4 null bytes and then provided with the subkey
+   * to the chacha20poly1305 implementation.
+   *
+   * If dst is given (it must be the size of plaintext + the size of tag
+   * length) the result will be put into it. Dst and plaintext must not
+   * overlap.
+   */
+  open(
+    nonce: Uint8Array,
+    sealed: Uint8Array,
+    associatedData?: Uint8Array,
+    dst?: Uint8Array
+  ): Uint8Array | null {
+    if (nonce.length !== 24) {
+      throw new Error("XChaCha20Poly1305: incorrect nonce length");
+    }
+
+    // Sealed ciphertext should at least contain tag.
+    if (sealed.length < this.tagLength) {
+      // TODO(dchest): should we throw here instead?
+      return null;
+    }
+
+    /**
+    * Generate subKey by using HChaCha20 function as defined
+    * in section 2 step 1 of draft-irtf-cfrg-xchacha-03
+    */
+    const subKey = hchacha(
+      this._key,
+      nonce.subarray(0, 16),
+      new Uint8Array(32)
+    );
+
+    /**
+    * Generate Nonce as defined - remaining 8 bytes of the nonce prefixed with
+    * 4 NUL bytes
+    */
+    const modifiedNonce = new Uint8Array(12);
+    modifiedNonce.set(nonce.subarray(16), 4);
+
+    // Decrypt.
+    const chachaPoly1305 = new ChaCha20Poly1305(subKey);
+
+    // Allocate space for decrypted plaintext.
+    const resultLength = sealed.length - this.tagLength;
+    let result;
+    if (dst) {
+      if (dst.length !== resultLength) {
+        throw new Error("XChaCha20Poly1305: incorrect destination length");
+      }
+      result = dst;
+    } else {
+      result = new Uint8Array(resultLength);
+    }
+
+    /**
+     * decrypt by calling into chacha20poly1305 as is described should be done
+     * in draft-
+     */
+    chachaPoly1305.open(modifiedNonce, sealed, associatedData, result);
```

Comment:
`open` returns `null` if authentication fails. Need to check for that and also return `null` otherwise, the result will be equal to destination or new allocated array instead of null:

```
if (!chachaPoly1305.open(...)) {
    return null;
}
```

But since we also can leave it to chachaPoly1305 to deal with allocating destination, we can just do:

```
const chachaPoly1305 = new ChaCha20Poly1305(subKey);
const result = chachaPoly1305.open(modifiedNonce, sealed, associatedData, dst);
wipe(modifiedNonce);
return result;
```

Reply:
I was curious what your opinion was on this. I found it somewhat difficult to figure out exactly how I should be calling into `XChacha20` and generating the tag because the RFC defines the usage of `AEAD_Chacha20_poly1305`. Do you think it would be more useful to call directly to `XChacha20` in this function rather than call to `Chacha20poly1305`?